### PR TITLE
feat: added setup for payment method management widget

### PR DESCRIPTION
**Pull request by Shivam25092001**
## ADDED
- Payment Method Management Widget

## REMOVED
- Payment Method Management FullScreen View

## SCREENSHOTS

Screenshot 2024-10-18 at 5 52 37 PM
Screenshot 2024-10-18 at 5 52 56 PM
Screenshot 2024-10-18 at 5 55 29 PM

## RECORDING

https://github.com/user-attachments/assets/4794cb1b-1da3-4da5-b8a8-cf97127bf469

## STACKED PR

- #14

### CODE REVIEW

File: hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagement.swift [hunk 1]
```diff
@@ -5,89 +5,89 @@
...
-            else if (response == "cancelled"){
-                completion(.closed(data: "cancelled"))
-            }
-            else {
-                completion(.closed(data: response ?? "failed"))
-            }
-        }
-    }
-}
+//import Foundation
```
Comment (25harsh) [inline hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagement.swift:8]: why is all this commented out, replace this code with widget code

File: hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagementWidget.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        
+        // Add the React Native view to the current view.
+        addSubview(rootView)
+    }
+    
+    internal static func exitWidget() {
+        print("-- exit widget")
+        DispatchQueue.main.async {
+            if let view = RNViewManager.sharedInstance.rootView {
+                let reactNativeVC: UIViewController? = view.reactViewController()
```
Comment (25harsh) [inline hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagementWidget.swift:85]: you are adding widget view to a UIViewController and then dismissing it, try to just remove the view

**Review by 25harsh — CHANGES_REQUESTED** [review]

File: hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift [hunk 3]
```diff
@@ -33,11 +33,26 @@ extension PaymentSession {
         paymentSheet.present(from: viewController, completion: completion)
     }
     
-    public func presentPaymentManagementSheet(viewController: UIViewController, configuration: PMMConfiguration, completion: @escaping (PaymentMethodManagementSheetResult) -> ()) {
-        let pmManagement = PaymentMethodManagement(ephemeralKey: PaymentSession.ephemeralKey ?? "", configuration: configuration)
-        pmManagement.presentPaymentMethodManagementView(from: viewController, completion: completion)
+    public func getPaymentMethodManagementWidget(frame: CGRect, onAddPaymentMethod: (() -> Void)?, completion: @escaping (PaymentMethodManagementResult) -> ()) -> UIControl {
```
[the remark this answers is not in the document]
Comment (Shivam25092001) [reply hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift:36]: but that would become asymetric to how we consume the paymentSheet. If that's ok then this is just a wrapper function, we can eliminate that

File: hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagement.swift (hunk 1 above)
Comment (Shivam25092001) [reply hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagement.swift:8]: we'll remove this file

File: hyperswitch/AddPaymentMethodViewController.swift [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+//  AddPaymentMethodViewController.swift
+//  hyperswitch
+//
+//  Created by Shivam Nan on 18/10/24.
+//
+
+import Foundation
+import UIKit
+
+class AddPaymentMethodViewController: UIViewController {
```
[the remark this answers is not in the document]
Comment (Shivam25092001) [reply hyperswitch/AddPaymentMethodViewController.swift:11]: added for now to depict the flow, I'll remove this in next PR once we start consuming the payment sheet

File: hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (25harsh) [reply hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift:36]: correct this is different from payment sheet, lets eliminate the wrapper function

File: hyperswitch/AddPaymentMethodViewController.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (25harsh) [reply hyperswitch/AddPaymentMethodViewController.swift:11]: sure

File: hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shivam25092001) [reply hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift]: done 👍

File: hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift [hunk 5]
```diff
@@ -33,10 +33,10 @@ extension PaymentSession {
         paymentSheet.present(from: viewController, completion: completion)
     }
     
-    public func presentPaymentManagementSheet(viewController: UIViewController, configuration: PMMConfiguration, completion: @escaping (PaymentMethodManagementSheetResult) -> ()) {
-        let pmManagement = PaymentMethodManagement(ephemeralKey: PaymentSession.ephemeralKey ?? "", configuration: configuration)
-        pmManagement.presentPaymentMethodManagementView(from: viewController, completion: completion)
-    }
+//    public func presentPaymentManagementSheet(viewController: UIViewController, configuration: PMMConfiguration, completion: @escaping (PaymentMethodManagementSheetResult) -> ()) {
```
Comment (25harsh) [inline hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift:36]: remove

File: hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagementWidget.swift [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    
+    public override init(
+        frame: CGRect
+    ) {
+        super.init(frame: frame)
+        commonInit()
+    }
+    
+    //    public override var intrinsicContentSize: CGSize {
```
Comment (25harsh) [inline hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagementWidget.swift:46]: remove

File: hyperswitch/PMMViewController.swift [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    @ObservedObject var hyperViewModel = HyperViewModel()
+    private var paymentSession: PaymentSession?
+    private var cancellables = Set<AnyCancellable>()
+    
+    private func setupPaymentWidget(topElement: UIView, onAddPaymentMethod: @escaping () -> Void) {
+        guard hyperViewModel.paymentSession != nil else { return }
+        
+        let paymentWidget = PaymentMethodManagementWidget(
+            ephemeralKey: PaymentSession.ephemeralKey ?? "",
+            frame: CGRect(x: 0, y: 0, width: view.bounds.width, height: view.bounds.height-160),
```
Comment (25harsh) [inline hyperswitch/PMMViewController.swift:16]: is this needed, does it actually change the size?

**Review by 25harsh — CHANGES_REQUESTED** [review]

File: hyperswitch/PMMViewController.swift (hunk 7 above)
Comment (Shivam25092001) [reply hyperswitch/PMMViewController.swift:16]: it does, it's required to define the view's bounds

Comment (25harsh) [reply hyperswitch/PMMViewController.swift:16]: what happens if you give 1/2 height of VC, can you add a screenshot

Comment (Shivam25092001) [reply hyperswitch/PMMViewController.swift:16]: image

_only white part of the screen is widget, while the blue part is merchants' view._

PMM widget's frame defined as:
```
let paymentWidget = PaymentMethodManagementWidget(
            frame: CGRect(x: 0, y: 0, width: view.bounds.width, **height: view.bounds.height/2**),
            ...
            )
```

File: hyperswitch/PMMViewController.swift [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    
+    private func asyncBindPaymentManagementWidget(topElement: UIView, onAddPaymentMethod: @escaping () -> Void) {
+        hyperViewModel.$status
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] status in
+                switch status {
+                case .loading:
+                    print("Loading payment method management session...")
+                case .success:
+                    self?.setupPaymentWidget(topElement: topElement, onAddPaymentMethod: onAddPaymentMethod)
```
[the remark this answers is not in the document]
Comment (Shivam25092001) [reply hyperswitch/PMMViewController.swift]: done 👍

File: hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagementWidget.swift (hunk 6 above)
Comment (Shivam25092001) [reply hyperswitchSDK/Core/HyperPaymentMethodManagement/PaymentMethodManagementWidget.swift]: done 👍

File: hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift (hunk 5 above)
Comment (Shivam25092001) [reply hyperswitchSDK/Core/HyperSession/PaymentSession+UIKit.swift]: done 👍

File: hyperswitch/HyperViewModel.swift [hunk 9]
```diff
@@ -21,25 +21,25 @@ class HyperViewModel: ObservableObject {
     }
     
     private func fetchData(from endpoint: String) async throws -> [String: Any] {
-            guard let url = URL(string: endpoint, relativeTo: backendUrl) else {
```
[the remark this answers is not in the document]
Comment (Shivam25092001) [reply hyperswitch/HyperViewModel.swift:24]: this fixes indentation

File: hyperswitch/PMMViewController.swift [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    @ObservedObject var hyperViewModel = HyperViewModel()
+    private var paymentSession: PaymentSession?
+    private var cancellables = Set<AnyCancellable>()
+    private var topElement: UIView?
+    
+    private func setupPaymentWidget(onAddPaymentMethod: @escaping () -> Void) {
+        guard hyperViewModel.paymentSession != nil else { return }
+        
+        let paymentWidget = PaymentMethodManagementWidget(
+            ephemeralKey: PaymentSession.ephemeralKey ?? "",
```
Comment (25harsh) [inline hyperswitch/PMMViewController.swift:16]: why accessing PaymentSession.ephemeralKey from internal code, use the model to get this

File: hyperswitch/PMMViewController.swift [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        NSLayoutConstraint.activate([
+            backButton.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
+            backButton.centerYAnchor.constraint(equalTo: titleLabel.centerYAnchor),
+            titleLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            titleLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20)
+        ])
+        
+        
+        let onAddPaymentMethod: () -> Void = {
+            DispatchQueue.main.async {
```
Comment (25harsh) [inline hyperswitch/PMMViewController.swift:90]: not needed for non ui changes

File: hyperswitch/PMMViewController.swift [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                configuration.displaySavedPaymentMethods = false
+                
+                var appearance = PaymentSheet.Appearance()
+                appearance.colors.background = UIColor(red: 0.96, green: 0.97, blue: 0.98, alpha: 1.00)
+                appearance.primaryButton.cornerRadius = 32
+                configuration.appearance = appearance
+
+                self.hyperViewModel.paymentSession?.presentPaymentSheet(viewController: self, configuration: configuration, completion: { result in
+                    DispatchQueue.main.async {
+                        print(result)
```
Comment (25harsh) [inline hyperswitch/PMMViewController.swift:103]: remove

Comment (Shivam25092001) [reply hyperswitch/PMMViewController.swift]: rather added a logic to refresh the list after successful completion

File: hyperswitch/PMMViewController.swift (hunk 10 above)
Comment (Shivam25092001) [reply hyperswitch/PMMViewController.swift]: removed ephemeral key from the init call param, as it was already defined while creating `PaymentSession`.

**Review by 25harsh — DISMISSED** [review]

**Review by 25harsh — APPROVED** [review]

**Pull request merged**